File: test/div10/div10.c

/* Tests if decimal output works, using division by 10 - GCC should optimize it so that no actual division is required */

#include <stddef.h>
#define ABI_IGNORE_ARGC_ARGV_ENVP 1
#include "syscall.h"

void putdec_u(size_t value)
{
	char buffer[sizeof(size_t) * 5 / 2];
	size_t ptr = sizeof buffer;
	do
	{
		int d = value % 10;
		value /= 10;
		buffer[--ptr] = '0' + d;
	} while(value != 0);
	write(1, &buffer[ptr], sizeof buffer - ptr);
}

int main(void)
{
	putdec_u(12345);
	return 123;
}

File: test/div10/div10.java

/* Tests if decimal output works, using division by 10 */

package div10;

import abi.Linux;

public class div10 {
	public static void main(String args[]) {
		// include the native library
		System.load(System.getProperty("user.dir") + "/syscall.so");

		// transfer to low-level interface
		_start();
	}

	static void _start() {
		Linux.exit(main());
	}

	static void putdec_u(int value) {
		int current_brk_value = Linux.brk(-1);
		byte[] buffer = new byte[4 * 5 / 2];
		int ptr = 4 * 5 / 2;
		do
		{
			int d = value % 10;
			value /= 10;
			buffer[--ptr] = (byte)('0' + d);
		} while(value != 0);
		Linux.write(1, buffer, ptr, 4 * 5 / 2 - ptr);
		Linux.brk(current_brk_value);
	}

	static int main() {
		putdec_u(12345);
		return 0;
	}
}

